torchtune/modules/transformer.py

```diff
@@ -4,42 +4,80 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 import copy
-from typing import Optional
+from abc import abstractmethod
+from typing import Optional, Protocol, runtime_checkable
 
 import torch
 from torch import nn, Tensor
 
-from torchtune.modules import CausalSelfAttention, KVCache
+from torchtune.modules import GroupedQueryAttention, KVCache
 
 
-class TransformerDecoderLayer(nn.Module):
+@runtime_checkable
```

Comment:
At least for this case I'm not sure it's all that helpful. Per [this comment](https://discuss.python.org/t/is-there-a-downside-to-typing-runtime-checkable/20731/2) I think it'd just be checking that forward is defined, and not actually checking the args and their types (which I think is the goal here)

Reply:
It was meant to enforce a consistent forward signature for layers, but I remove it. The benefit of this would be a form of self documentation for people making custom transformer layers.